meter/shelly/gen2.go

```diff
@@ -51,32 +63,239 @@ type Gen2EM1Data struct {
+	// Shelly gen 2 rfc7616 authentication
+	// https://shelly-api-docs.shelly.cloud/gen2/Overview/CommonDeviceTraits#authentication
+	// https://datatracker.ietf.org/doc/html/rfc7616
+
+	data := &Gen2RpcPost{
+		Id:     c.channel,
+		On:     enable,
+		Src:    "evcc",
+		Method: method,
+	}
+
+	req, err := request.New(http.MethodPost, fmt.Sprintf("%s/%s", c.uri, method), request.MarshalJSON(data), request.JSONEncoding)
+	if err != nil {
+		return err
+	}
+
+	return c.DoJSON(req, &res)
+}
+
+// CurrentPower implements the api.Meter interface
+func (c *gen2) CurrentPower() (float64, error) {
+	switch {
+	// Endpoint Switch.GetStatus
+	case c.hasSwitchEndpoint():
+		res, err := c.switchstatus.Get()
+		return res.Apower, err
+
+		// Endpoint EM1.GetStatus
+	case c.hasEMEndpoint() && c.profile == "monophase":
+		res, err := c.em1status()
+		return res.ActPower, err
+
+		// Endpoint EM.GetStatus
+	case c.hasEMEndpoint() && c.profile == "triphase":
+		res, err := c.emstatus()
+		return res.TotalActPower, err
+
+	default:
+		return 0, fmt.Errorf("unknown shelly model: %s", c.model)
+	}
+}
+
+// Gen2Enabled implements the Gen2 api.Charger interface
+func (c *gen2) Enabled() (bool, error) {
+	if c.hasSwitchEndpoint() {
+		res, err := c.switchstatus.Get()
+		return res.Output, err
+	}
+
+	return false, fmt.Errorf("unknown shelly model: %s", c.model)
+}
+
+// Gen2Enable implements the api.Charger interface
+func (c *gen2) Enable(enable bool) error {
+	var res Gen2SwitchStatus
+	c.switchstatus.Reset()
+	return c.execCmd("Switch.Set?id="+strconv.Itoa(c.channel), enable, &res)
+}
+
+// TotalEnergy implements the api.Meter interface
+func (c *gen2) TotalEnergy() (float64, error) {
+	switch {
+	case c.hasSwitchEndpoint():
+		res, err := c.switchstatus.Get()
+		return res.Aenergy.Total / 1000, err
+
+	case c.hasEMEndpoint() && c.profile == "monophase":
+		res, err := c.em1data()
+		return res.TotalActEnergy / 1000, err
+
+	case c.hasEMEndpoint() && c.profile == "triphase":
+		res, err := c.emdata()
+		return res.TotalAct / 1000, err
+
+	default:
+		return 0, fmt.Errorf("unknown shelly model: %s", c.model)
+	}
+}
+
+// Currents implements the api.PhaseCurrents interface
+func (c *gen2) Currents() (float64, float64, float64, error) {
+	switch {
+	case c.hasSwitchEndpoint():
+		res, err := c.switchstatus.Get()
+		return res.Current, 0, 0, err
+
+	case c.hasEMEndpoint() && c.profile == "monophase":
+		res, err := c.em1status()
+		return res.Current, 0, 0, err
+
+	case c.hasEMEndpoint() && c.profile == "triphase":
+		res, err := c.emstatus()
+		return res.ACurrent, res.BCurrent, res.CCurrent, err
+
+	default:
+		return 0, 0, 0, fmt.Errorf("unknown shelly model: %s", c.model)
+	}
+}
+
+// Voltages implements the api.PhaseVoltages interface
+func (c *gen2) Voltages() (float64, float64, float64, error) {
+	switch {
+	case c.hasSwitchEndpoint():
+		res, err := c.switchstatus.Get()
+		return res.Voltage, 0, 0, err
```

Comment:
Nur laut gedacht und ohne, dass wir das machen müssten: wenn man jetzt den Status Typen auch ein Interface für Vortage, Current, Energy verpasst dann kann man den gen2 mit dem Status Typ parametrieren und die ganzen switch/case Statements fallen weg und es verbleibt ein einziger Cache für die Statusabfrage.

Reply:
kann ich mir anschauenm ist aber in der Regel eine Kombi aus Status und Data Endpoints Ergebnis und andere Power Quellfelder aus den Response Structs.